File: Uva/11624 - Fire/src/Main.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;




public class Main {
	static Scanner sc =  new Scanner(new InputStreamReader(System.in));
	static char  map[][];
	static int vif[][];
	static int vij[][];
	static int dx[]= {0,0,1,-1};
	static int dy[] = {1,-1,0,0};
	static Queue<Par> qj ;
	static Queue<Par> qf;
	static int r , c  ,  xj  , yj;
	public static void main(String[] args) throws Exception{	
	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	int casos =  Integer.parseInt(br.readLine().trim());
	String datos[];
	String line;
	boolean jenc =  false;
	for(int i = 0 ;  i <  casos ;  i++){
		
		datos =  br.readLine().trim().split(" ");
		
		r =  Integer.parseInt(datos[0]);
		c =  Integer.parseInt(datos[1]);
		
		map =  new char[r][c];
		vif = new int[r][c];
		vij =  new int[r][c];
		qj =  new LinkedList<Par>();
		qf =  new LinkedList<Par>();
		
		for(int  m=  0 ;  m <  r ; m++){
			for(int n = 0 ;  n  < c ; n++ ){
				vif[m][n] =  Integer.MAX_VALUE;
				vij[m][n] =  Integer.MAX_VALUE;
			}
		}
		
		for(int j = 0 ;  j < r ; j++){
			line = br.readLine().trim();
			
			for(int k = 0  ;  k  < c ;  k++){
				if( !jenc && line.charAt(k) == 'J' ){
					xj =  j;
					yj =  k;
				}
				
				
				if( line.charAt(k) ==  'F'){
					qf.add(new Par(j  , k));
					vif[j][k] = 0;
				}
				
				map[j][k] = line.charAt(k);
			}
		}
		
		
		int res =  solve();
		if(res == Integer.MAX_VALUE) System.out.println("IMPOSSIBLE");
		else System.out.println(res + 1);
		
		
	}
	}
	private static int solve() {
		
		
	
		
		Par inicio = new Par(xj , yj);
		
		qj.add(inicio);
		vij[xj][yj] = 0;
		
		
		//recorrer el jugador
		while(!qj.isEmpty()){
			Par u =  qj.poll();
			for(int i = 0 ;  i <  4 ;  i++){
				int nx =  u.x + dx[i];
				int ny =  u.y + dy[i];
				
				if( nx >= 0  && nx < r && ny >= 0 && ny < c &&  map[nx][ny]!= '#' &&  vij[nx][ny] == Integer.MAX_VALUE ){
					qj.add(new Par(nx , ny));
					vij[nx][ny] =  vij[u.x][u.y] + 1;
				}
			}
		}
		
		
		
		while(!qf.isEmpty()){
			Par u =  qf.poll();
			for(int i = 0 ;  i <  4 ;  i++){
				int nx =  u.x + dx[i];
				int ny =  u.y + dy[i];
				if( nx >= 0  && nx < r && ny >= 0 && ny < c && map[nx][ny]!= '#' &&   vif[nx][ny] == Integer.MAX_VALUE  ){
					qf.add(new Par(nx , ny));
					vif[nx][ny] =  vif[u.x][u.y] + 1;
				}
			}
		}
		
		int minf = Integer.MAX_VALUE;
		for(int  i = 0 ; i < c ;  i++){
			if( vif[0][i] > vij[0][i] ){
				minf = Math.min(minf, vij[0][i]);
			}
			if( vif[r- 1][i] > vij[r - 1][i] ){
				minf = Math.min(minf, vij[r - 1][i]);
			}
		}
		
		for(int  i = 0 ; i < r ;  i++){
			if( vif[i][0] > vij[i][0] ){
				minf = Math.min(minf, vij[i][0]);
			}
			if( vif[i][c - 1] > vij[i][c - 1] ){
				minf = Math.min(minf, vij[i][c - 1]);
			}
		}
		
		return minf;
	}

}

class Par{
	int x ,  y;
	public Par(int x, int y){
		this.x = x;
		this.y = y;
	}
}

File: Uva/11624 - Fire/main.cpp




#include <cstdio>
#include <algorithm>
#include <iostream>
#define FOR(i, a, b)        for(i=a; i<b; i++)
#define INF_MAX 2147483647
#include <queue>
using namespace std;
    char mapa[1005][1005];
    int vif[1005][1005];
    int vij[1005][1005];
int dx[] = {0,0,1,-1};
int dy[] = {1,-1,0,0};
int r , c;

struct Par{
    int i ,  j;
    Par( int ii ,  int jj ) { i = ii ; j = jj;}
};


void bfs_joe( int i , int j){
Par inicio(i , j);
vij[i][j] = 0;
queue<Par> qj;
qj.push(inicio);

while(!qj.empty()){
    int i ;
    int nx ,  ny;

    Par u  = qj.front();
    qj.pop();
    FOR( i , 0 , 4){
        nx = u.i + dx[i];
        ny = u.j + dy[i];

        if( nx  >=  0 &&  nx <  r && ny >= 0 &&  ny < c && vij[nx][ny] == INF_MAX && mapa[nx][ny] != '#'){
            vij[nx][ny] = vij[u.i][u.j] + 1;
            qj.push(Par(nx , ny));
        }
    }
}


}

int main(){
    int casos;

    scanf("%d" ,  &casos);


    while(casos--){


    scanf("%d  %d" ,  &r , &c);

    queue<Par> qf;


    int i , j  ,  xj , yj;
    FOR(i ,  0 , r )
        FOR(j , 0 , c){
            vif[i][j]  =INF_MAX;
            vij[i][j]  = INF_MAX;
        }

    FOR(i , 0 , r){
        scanf("%s"  , mapa[i]);
         FOR( j  , 0 ,  c){
            if(mapa[i][j] == 'J'){
              xj = i ; yj = j;
            }
            if(mapa[i][j] == 'F'){
                Par a(i , j);
                qf.push(a);
                vif[i][j] = 0;
            }
        }

    }

    //Hacer el bfs para joe
    bfs_joe(xj ,  yj);

    //
    int res;



    while(!qf.empty()){

        Par u = qf.front();
        qf.pop();
        int i ;
        FOR( i , 0 ,  4){
            int nx =  u.i + dx[i];
            int ny =  u.j + dy[i];
				if( nx >= 0  && nx < r && ny >= 0 && ny < c && mapa[nx][ny]!= '#' &&   vif[nx][ny] == INF_MAX ){
					qf.push( Par(nx , ny));
					vif[nx][ny] =  vif[u.i][u.j] + 1;
				}
        }
    }




		int minf = INF_MAX;
		for(int  i = 0 ; i < c ;  i++){
			if( vif[0][i] > vij[0][i] ){
				minf = min(minf, vij[0][i]);
			}
			if( vif[r - 1][i] > vij[r - 1][i] ){
				minf = min(minf, vij[r - 1][i]);
			}
		}

		for(int  i = 0 ; i < r ;  i++){
			if( vif[i][0] > vij[i][0] ){
				minf = min(minf, vij[i][0]);
			}
			if( vif[i][c - 1] > vij[i][c - 1] ){
				minf = min(minf, vij[i][c - 1]);
			}
		}


        if(minf == INF_MAX) printf("IMPOSSIBLE\n");
        else printf("%d\n"   ,  (minf + 1));








    }

    return  0;

}


